File: Crio/crio_programming_interview_problems-master/BuildOrder/BuildOrder.cpp
#include<bits/stdc++.h>
using namespace std;
vector<string> buildOrder(vector<string>& project, vector<vector<string>>& depend){
    // CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
    int n = project.size();
    map<string,int>hash;
    map<int,string>hashrev;
    int c = 1;
    for (auto &&name : project)
    {
        hashrev[c] = name;
        hash[name] = c++;
    }
    vector<int>inD(n+1);
    vector<vector<int>>adj(n+1);
    for (auto &&edge : depend)
    {
        inD[hash[edge[1]]]++;
        adj[hash[edge[0]]].push_back(hash[edge[1]]);
    }
    queue<int>q;
    vector<string>ans;
    for (int i = 1; i <= n; i++)
    {
        if(inD[i]==0){
            q.push(i);
        }
    }
    while(!q.empty()){
        int p = q.front();
        q.pop();
        ans.push_back(hashrev[p]);
        for (auto &&child : adj[p])
        {
            inD[child]--;
            if(inD[child]==0){
                q.push(child);
            }
        }
    }
    if(ans.size()!=n){
        ans.clear();
        ans.push_back("-1");
    }
    return ans;
    // CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
}
int main(){

    int n,d;
    cin>>n>>d;
    vector<string>project(n);
    vector<vector<string>>depend(d);
    for (int i = 0; i < n; i++)
    {
        cin>>project[i];
    }

    for (int i = 0; i < d; i++)
    {
        string u,v;
        cin>>u>>v;
        depend[i].push_back(u);
        depend[i].push_back(v);
    }
    vector<string> ans = buildOrder(project,depend);
    for (auto p : ans)
    {
        cout<<p<<" ";
    }
    
}

File: Crio/crio_programming_interview_problems-master/BuildOrder/BuildOrder.java
import java.util.*;

public class BuildOrder {
    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);

        int p = sc.nextInt();
        int d = sc.nextInt();
        
        sc.nextLine();
        ArrayList<String> project = new ArrayList<>();
        for(String val:sc.nextLine().split(" ")){
            // System.out.print(val+" ");
            project.add(val);
        }
        
        ArrayList<ArrayList<String> > dependencies = new ArrayList<>();
        for(int i=0;i<d;i++){
            ArrayList<String> dependency = new ArrayList<>();
            for(String dep:sc.nextLine().split(" ")){
                dependency.add(dep);
            }
            dependencies.add(dependency);
        }
        ArrayList<String> ans = buildOrder(project,dependencies);

        for(String proj:ans){
            System.out.print(proj+" ");
        }
        sc.close();
    }
    public static ArrayList<String> buildOrder(ArrayList<String> project,ArrayList<ArrayList<String> > dependencies){
// CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
	    
        int n = project.size();
        ArrayList<ArrayList<Integer> > adj = new ArrayList<>();
        ArrayList<String> ans = new ArrayList<>();
        HashMap<String,Integer> projectToIndex = new HashMap<>();
        HashMap<Integer,String> indexToProject = new HashMap<>();
        ArrayList<Integer> deg = new ArrayList<>();
        for(int i =0;i<n;i++){
            projectToIndex.put(project.get(i),i);
            indexToProject.put(i,project.get(i));
            deg.add(0);
            adj.add(new ArrayList<>());
        }
        
        for(ArrayList<String> dependency:dependencies){
            int val = deg.get(projectToIndex.get(dependency.get(1)));
            deg.set(projectToIndex.get(dependency.get(1)),val+1);
            adj.get(projectToIndex.get(dependency.get(0))).add(projectToIndex.get(dependency.get(1)));
        }

        Queue<Integer> q = new LinkedList<>();

        for(int i=0;i<n;i++){
            if(deg.get(i)==0){
                q.add(i);
            }
        }

        while(!q.isEmpty()){
            int top = q.remove();
            ans.add(indexToProject.get(top));

            for(int child:adj.get(top)){
                deg.set(child,deg.get(child)-1);
                if(deg.get(child)==0){
                    q.add(child);
                }
            }
        }

        if(ans.size()<n){
            ans.clear();
            ans.add("-1");
        }

        
        return ans;
// CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
    }

}
